src/compute-client/src/controller/introspection.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use mz_repr::{Diff, Row};
+use mz_storage_client::client::AppendOnlyUpdate;
+use mz_storage_client::controller::{IntrospectionType, StorageController, StorageWriteOp};
+use mz_storage_types::controller::StorageError;
+use timely::progress::Timestamp;
+use tokio::sync::{mpsc, oneshot};
+use tracing::info;
+
+pub type IntrospectionUpdates = (IntrospectionType, Vec<(Row, Diff)>);
+
+/// Spawn a task sinking introspection updates produced by the compute controller to storage.
+pub fn spawn_introspection_sink<T: Timestamp>(
+    mut rx: mpsc::UnboundedReceiver<IntrospectionUpdates>,
+    storage_controller: &dyn StorageController<Timestamp = T>,
+) {
+    let sink = IntrospectionSink::new(storage_controller);
+
+    mz_ore::task::spawn(|| "compute-introspection-sink", async move {
+        info!("running introspection sink task");
+
+        while let Some((type_, updates)) = rx.recv().await {
```

Comment:
Would it make sense to `recv_many` here, consolidate, and send a batch of updates?

Reply:
I was considering that but decided against it because it would require collecting updates by `type_` here, which is a bit annoying. I think it wouldn't help too much because the individual `updates` batches should already be consolidated given how the producers work. It would help if we could merge multiple `updates` batches for the same time, but I wouldn't expect to encounter that often, given that this task immediately forwards batches and doesn't wait for them to accumulate.